File: experiments/gilded-rose/fsharp/attempts/opus46high/1/GildedRose.fs
module GildedRose

open System.Collections.Generic

type Item = { Name: string; SellIn: int; Quality: int }

type private ItemCategory =
    | AgedBrie
    | BackstagePass
    | Sulfuras
    | Conjured
    | Normal

let private categorize (name: string) =
    match name with
    | "Aged Brie" -> AgedBrie
    | "Backstage passes to a TAFKAL80ETC concert" -> BackstagePass
    | "Sulfuras, Hand of Ragnaros" -> Sulfuras
    | n when n.StartsWith("Conjured") -> Conjured
    | _ -> Normal

let private clampQuality quality = max 0 (min 50 quality)

let private updateItem (item: Item) =
    match categorize item.Name with
    | Sulfuras -> item
    | AgedBrie ->
        let increase = if item.SellIn <= 0 then 2 else 1
        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality (item.Quality + increase) }
    | BackstagePass ->
        let quality =
            if item.SellIn <= 0 then 0
            elif item.SellIn <= 5 then item.Quality + 3
            elif item.SellIn <= 10 then item.Quality + 2
            else item.Quality + 1
        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality quality }
    | Conjured ->
        let degradation = if item.SellIn <= 0 then 4 else 2
        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality (item.Quality - degradation) }
    | Normal ->
        let degradation = if item.SellIn <= 0 then 2 else 1
        { item with
            SellIn = item.SellIn - 1
            Quality = clampQuality (item.Quality - degradation) }

type GildedRoseShop(items: IList<Item>) =
    member _.Items = items

    member _.UpdateQuality() =
        for i = 0 to items.Count - 1 do
            items.[i] <- updateItem items.[i]

File: experiments/gilded-rose/elixir/attempts/opus46high/1/lib/gilded_rose.ex
defmodule GildedRose do
  def update_quality(items) do
    Enum.map(items, &update_item/1)
  end

  def update_item(%Item{name: "Sulfuras, Hand of Ragnaros"} = item), do: item

  def update_item(%Item{name: "Aged Brie"} = item) do
    item
    |> decrease_sell_in()
    |> increase_quality(if(item.sell_in <= 0, do: 2, else: 1))
  end

  def update_item(%Item{name: "Backstage passes to a TAFKAL80ETC concert"} = item) do
    item
    |> decrease_sell_in()
    |> update_backstage_quality()
  end

  def update_item(%Item{name: "Conjured " <> _} = item) do
    item
    |> decrease_sell_in()
    |> decrease_quality(if(item.sell_in <= 0, do: 4, else: 2))
  end

  def update_item(item) do
    item
    |> decrease_sell_in()
    |> decrease_quality(if(item.sell_in <= 0, do: 2, else: 1))
  end

  defp decrease_sell_in(item), do: %{item | sell_in: item.sell_in - 1}

  defp increase_quality(item, amount) do
    %{item | quality: min(item.quality + amount, 50)}
  end

  defp decrease_quality(item, amount) do
    %{item | quality: max(item.quality - amount, 0)}
  end

  defp update_backstage_quality(%Item{sell_in: sell_in} = item) when sell_in < 0 do
    %{item | quality: 0}
  end

  defp update_backstage_quality(%Item{sell_in: sell_in} = item) when sell_in < 5 do
    increase_quality(item, 3)
  end

  defp update_backstage_quality(%Item{sell_in: sell_in} = item) when sell_in < 10 do
    increase_quality(item, 2)
  end

  defp update_backstage_quality(item) do
    increase_quality(item, 1)
  end
end
